tests/liteclientHSM_test.cc

```diff
@@ -0,0 +1,1179 @@
+    cmd = boost::str(doCsr % cnf_ % key);
+    if (Utils::shell(cmd, &out, true) != EXIT_SUCCESS) {
+      LOG_INFO << "Error: " << out;
+      throw std::runtime_error(cmd.c_str());
+    }
+
+    /* write CSR to disk */
+    csr = hsm_->path_ + csr;
+    Utils::writeFile(csr, out, true);
+  }
+
+  void createCrt(std::string csr, std::string& crt) {
+    crt = hsm_->path_ + crt;
+    rootCa_.signCsr(csr, crt, "");
+  }
+
+  void importCrt(std::string& crt, std::string id) {
+    boost::format crtToDer("OPENSSL_CONF=%s openssl x509 -inform pem -in %s -out %s/tmp.der");
+    cmd = boost::str(crtToDer % cnf_ % crt % hsm_->path_);
+    if (Utils::shell(cmd, &out, true) != EXIT_SUCCESS) {
+      LOG_INFO << "Error: " << out;
+      throw std::runtime_error(cmd.c_str());
+    }
+
+    boost::format writeCrtToHsm("pkcs11-tool --module %s -w %s/tmp.der -y cert --id %s --pin %s");
+    cmd = boost::str(writeCrtToHsm % hsm_->module_ % hsm_->path_ % id % hsm_->pin_);
+    if (Utils::shell(cmd, &out, true) != EXIT_SUCCESS) {
+      LOG_INFO << "Error: " << out;
+      throw std::runtime_error(cmd.c_str());
+    }
+  }
+
+  void listInfo() {
+    boost::format listMechanisms("pkcs11-tool --module %s --list-mechanisms");
+    cmd = boost::str(listMechanisms % hsm_->module_);
+    if (Utils::shell(cmd, &out, true) != EXIT_SUCCESS) {
+      LOG_INFO << "Error: " << out;
+      throw std::runtime_error(cmd.c_str());
+    }
+    // very verbose
+    // LOG_INFO << out;
+    boost::format listObjects("pkcs11-tool --module %s --list-objects");
+    cmd = boost::str(listObjects % hsm_->module_);
+    if (Utils::shell(cmd, &out, true) != EXIT_SUCCESS) {
+      throw std::runtime_error(cmd.c_str());
+    }
+    // very verbose
+    // LOG_INFO << out;
+  }
+
+ private:
+  Hsm* hsm_;
+  RootCaPKI& rootCa_;
+  std::string cnf_;
+  /* buffers */
+  std::string cmd;
+  std::string out;
+};
+
+class SubscriberPKI {
+ public:
+  SubscriberPKI(DeviceHsm deviceHsm, std::string keyId, std::string certId, std::string keyLabel, std::string csr,
+                std::string crt)
+      : keyId_(std::move(keyId)),
+        certId_(std::move(certId)),
+        keyLabel_(std::move(keyLabel)),
+        csr_{std::move(csr)},
+        crt_{std::move(crt)} {
+    deviceHsm.createKey(keyId_, keyLabel_);
+    deviceHsm.createCsr(keyLabel_, csr_);
+    deviceHsm.createCrt(csr_, crt_);
+    deviceHsm.importCrt(crt_, certId_);
+    /* enable for debug */
+    deviceHsm.listInfo();
+  }
+
+ public:
+  std::string keyId_;
+  std::string certId_;
+
+ private:
+  std::string keyLabel_;
+  std::string csr_;
+  std::string crt_;
+};
+
+/**
+ * Class LiteClientHSMTest
+ *
+ */
+class LiteClientHSMTest : public ::testing::Test {
+ public:
+  static std::string SysRootSrc;
+
+ protected:
+  static void SetUpTestSuite() {
+    boost::filesystem::path path = boost::filesystem::temp_directory_path() / boost::filesystem::unique_path();
```

Comment:
Maybe instead of `boost::filesystem::unique_path()` it makes sense to define this subfolder as `hsm` or `pki`, just makes debugging a bit easier since you clearly see where is a folder with PKI stuff.